src/engine/sparqlExpressions/DateExpressions.cpp

```diff
@@ -40,32 +40,21 @@ inline auto extractDay = [](std::optional<DateOrLargeYear> d) {
   return Id::makeFromInt(optionalDay.value());
 };
 
-inline auto extractHours = [](std::optional<DateOrLargeYear> d) {
-  if (!d.has_value() || d->getType() != DateOrLargeYear::Type::DateTime) {
+template <auto dateMember, auto makeId>
+inline auto extractTimeComponentImpl = [](std::optional<DateOrLargeYear> d) {
+  if (!d.has_value() || !d->isDate()) {
     return Id::makeUndefined();
   }
-  auto hours = d.value().getDate().getHour();
-  if (hours == -1 ) {
+  Date date = d.value().getDate();
+  if (!date.hasTime()) {
     return Id::makeUndefined();
   }
-  return Id::makeFromInt(hours);
+  return std::invoke(makeId, std::invoke(dateMember, date));
 };
 
-inline auto extractMinutes = [](std::optional<DateOrLargeYear> d) {
-  if (!d.has_value() || d->getType() != DateOrLargeYear::Type::DateTime) {
-    return Id::makeUndefined();
-  }
-  auto minutes = d.value().getDate().getMinute();
-  return Id::makeFromInt(minutes);
-};
-
-inline auto extractSeconds = [](std::optional<DateOrLargeYear> d) {
-  if (!d.has_value() || d->getType() != DateOrLargeYear::Type::DateTime) {
-    return Id::makeUndefined();
-  }
-  auto seconds = d.value().getDate().getSecond();
-  return Id::makeFromDouble(seconds);
-};
+inline auto extractHours = extractTimeComponentImpl<&Date::getHour, &Id::makeFromInt>;
+inline auto extractMinutes = extractTimeComponentImpl<&Date::getMinute, &Id::makeFromInt>;
+inline auto extractSeconds = extractTimeComponentImpl<&Date::getSecond, &Id::makeFromDouble>;
```

Comment:
Very nice now,
Our static analysis complains, that the lambdas are global variables that are not const. While this in practice should be no problem, Maybe you could start cleaning up our code by making all the lambdas in this file `constexpr auto <name> = ...` instead of `inline auto`.

Reply:
Done.